android/titanium/libv8-services.js

```diff
@@ -0,0 +1,288 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2009-2019 by Axway. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License.
+ * Please see the LICENSE included with this distribution for details.
+ */
+
+'use strict';
+
+const AndroidBuilder = require('../../build/lib/android');
+const Builder = require('../../build/lib/builder');
+const util = require('util');
+const exec = util.promisify(require('child_process').exec); // eslint-disable-line security/detect-child-process
+const crypto = require('crypto');
+const fs = require('fs-extra');
+const path = require('path');
+const request = require('request-promise-native');
+
+// Determine if we're running on a Windows machine.
+const isWindows = (process.platform === 'win32');
+
+/**
+ * Double quotes given path and escapes double quote characters in file/directory names.
+ * @param {String} filePath The path to be double quoted. Can be null/undefined.
+ * @return {String} Returns the double quoted path. Will be null/undefined if given argument is null/undefined.
+ */
+function quotePath(filePath) {
+	if (!filePath) {
+		return '""';
+	}
+	if (!isWindows) {
+		filePath = filePath.replace(/"/g, '\\"');
+	}
+	return `"${filePath}"`;
+}
+
+/**
+ * Reads all of the bytes of the given file and returns a hash using the given algorithm.
+ * @param {String} filePath Path to the file to read the bytes of and calculate a hash. Cannot be null.
+ * @param {String} hashingAlgorithm Hashing algorithm to use such as 'md5', 'sha256', etc. Cannot be null.
+ * @param {String} [hashEncoding]
+ * Optional argument indicating how to encode the returned hash result.
+ * Can be set to 'hex', 'base64, or 'ascii', 'utf8'. Defaults to 'hex' if argument is null or undefined.
+ */
+async function fetchFileHash(filePath, hashingAlgorithm, hashEncoding) {
+	return await new Promise((resolve, reject) => {
+		const hash = crypto.createHash(hashingAlgorithm);
+		const readStream = fs.createReadStream(filePath);
+		readStream.on('data', (data) => {
+			hash.update(data);
+		});
+		readStream.on('end', () => {
+			if (!hashEncoding) {
+				hashEncoding = 'hex';
+			}
+			resolve(hash.digest(hashEncoding));
+		});
+		readStream.on('error', reject);
+	});
+}
+
+/**
+ * Async loads the "titanium_mobile/android/package.json" file and returns it as a dictionary.
+ * @returns {Promise<Object>} Dictionary of the parsed JSON file if loaded successfully.
+ */
+async function loadPackageJson() {
+	const filePath = path.join(__dirname, '..', 'package.json');
+	const fileContent = await fs.readFile(filePath, 'utf8');
+	return JSON.parse(fileContent);
+}
+
+/**
+ * Does a transpile/polyfill/rollup of our "titanium_mobile/common/Resources" JS files.
+ * Will then generate a C++ header file providing a V8 snapshot of the rolled-up JS for fast startup times.
+ */
+async function createSnapshot() {
```

Comment:
This could be simplified a little more:
```JS
async function createSnapshot() {
	// Perform a transpile/polyfill/rollup of our common JS files.
	// This contains our "ti.main.js" which the app executes on startup before executing the "app.js" file.
	const rollupOutputDirPath = path.join(__dirname, 'build', 'outputs', 'ti-assets', 'Resources');
	const rollupOutputFilePath = path.join(rollupOutputDirPath, 'ti.main.js');
	await fs.ensureDir(rollupOutputDirPath);
	const mainBuilder = new Builder({ args: [ 'android' ] });
	const androidBuilder = new AndroidBuilder({});
	await mainBuilder.transpile('android', androidBuilder.babelOptions, rollupOutputFilePath);
	const rollupFileContent = (await fs.readFile(rollupOutputFilePath)).toString();

	// Create the C++ directory we'll be generating the snapshot header file to.
	const cppOutputDirPath = path.join(__dirname, '..', 'runtime', 'v8', 'generated');
	const v8SnapshotHeaderFilePath = path.join(cppOutputDirPath, 'V8Snapshots.h');
	await fs.ensureDir(cppOutputDirPath);

	// Requests our server to create snapshot of rolled-up "ti.main" in a C++ header file.
	try {
		const MAX_ATTEMPTS = 20; // Allow for two minutes.

		// Post rolled-up "ti.main" script to server and obtain a snapshot ID as a response.
		// We will send an HTTP request for the snapshot code later.
		console.log('Attempting to request snapshot...');
		const snapshotUrl = 'http://v8-snapshot.appcelerator.com';
		const packageJsonData = await loadPackageJson();
		const requestOptions = {
			body: {
				v8: packageJsonData.v8.version,
				script: rollupFileContent
			},
			json: true
		};

		// Request `snapshotId` from server.
		const snapshotId = await request.post(snapshotUrl, requestOptions);
		for (let attempts = 0; attempts <= MAX_ATTEMPTS; attempts++) {

			// Request generated snapshot from server, if done.
			const response = await request.get(`${snapshotUrl}/snapshot/${snapshotId}`, {
				simple: false,
				resolveWithFullResponse: true
			});

			// Validate status codes.
			if (response.statusCode === 200) {

				// Server has finished creating a C++ header file containing all V8 snapshots.
				// Write it to file and flag that we're done.
				console.log('Writing snapshot...');
				await fs.writeFile(v8SnapshotHeaderFilePath, response.body);

			} else if (statusCode === 202 && attempts < MAX_ATTEMPTS) {

				// Snapshot server is still building. We need to retry later.
				console.log('Waiting for snapshot generation...');
				await new Promise(resolve => setTimeout(resolve, 6000));
				continue;
			}

			// Give up if received an unexpected response.
			console.error('Could not generate snapshot, skipping...');
			break;
		}
	} catch(e) {
		console.error(`Failed to request snapshot: ${e}`);

		// Generate an empty C++ header if the above failed.
		// Note: The C++ build will fail if file is missing. This is because it is #included in our code.
		await fs.writeFile(v8SnapshotHeaderFilePath, '// Failed to build V8 snapshots. See build log.');
	}
}
```

Reply:
Thanks. I agree. I've implemented this... but with some alterations.

I've added a new environment variable `TI_SDK_BUILD_REQUIRES_V8_SNAPSHOTS`, which when set to `1`, will purposely fail the build if we were unable to generate the snapshot. I'm setting this in our `titanium_mobile/build` script. The idea being we don't want Jenkins to accidentally release an SDK build without V8 snapshot support for fast app startup times. But I think our fallback mechanism is okay for local Android Studio builds; especially when building offline.